refactored navbar styling and logo spacings

bdbch opened the pull request:
Minor adjustments on the navbar including less menu items (Matches, Heroes, Combos) for more focused navigation and moved all other links to the expand menu.

I also increased the left and right padding of the navbar in total and gave the logo a bit more spacing.

Before:
![image](https://user-images.githubusercontent.com/6538827/66292308-4a679280-e8e4-11e9-8ae0-d9d0158e50b5.png)

After:
![image](https://user-images.githubusercontent.com/6538827/66292323-52273700-e8e4-11e9-903a-bfed8e976ef9.png)

bdbch commented:
I also refactored the login button. It won't show the spinning indicator on slow connections but just hide the widget until we know the authorization state.

I also used the @material-ui/core/Button component.

Before:
![image](https://user-images.githubusercontent.com/6538827/66292931-c57d7880-e8e5-11e9-9cbf-378f02348435.png)

After:
![image](https://user-images.githubusercontent.com/6538827/66292967-d928df00-e8e5-11e9-820c-b3748c0dcfaf.png)

It now has the default ripple effect.

howardchung commented on src/components/Header/Header.jsx:
```diff
@@ -88,9 +88,12 @@ const LinkGroup = ({ navbarPages }) => {
     setAnchorEl(undefined);
   }, [anchorEl]);
 
+  const mainNavbarPages = useMemo(() => navbarPages.slice(0, 3), [navbarPages]);
```
Rather than slicing by index, I think we should add a property in each navbar object that indicates whether it's a main one or a drawer one, and render the drawer ones in a burger menu as described in https://github.com/odota/web/issues/2222

howardchung commented:
If you want to merge the login button changes now, you can revert the navbar changes and submit them in a later PR?

bdbch commented:
@howardchung sure thing. will do that asap in the evening today.

> Rather than slicing by index, I think we should add a property in each navbar object that indicates whether it's a main one or a drawer one, and render the drawer ones in a burger menu as described in

So we would render the burger menu on desktop too right?

bdbch commented:
@howardchung this is how it looks like now:
![image](https://user-images.githubusercontent.com/6538827/66434017-e6e37f00-ea21-11e9-8b79-688452614df7.png)

![image](https://user-images.githubusercontent.com/6538827/66434031-eea32380-ea21-11e9-9510-c0042069269f.png)

I guess we still need the User login/logout in the drawer?

bdbch commented:
The app bar is now also fixed so you can open the drawer wherever you are:
![image](https://user-images.githubusercontent.com/6538827/66434190-4a6dac80-ea22-11e9-9bb7-4df132dd6109.png)

bdbch commented on src/components/Header/Header.jsx:
```diff
@@ -139,105 +144,130 @@ const SettingsGroup = ({ children }) => {
+  const navigationState = useContext(navigationContext);
+  const small = useSelector(state => state.browser.greaterThan.small);
+  const user = useSelector(state => state.app.metadata.data.user);
+  const strings = useSelector(state => state.app.strings);
```
So just for clarity for people who are not used to write components with hooks yet I've decided to add an explanation here on what is going on.

```js
// Here I create a state for the Announce component
// It's the equivalent to this.state.Announce and this.setState({ Announce })
const [Announce, setAnnounce] = useState(null);

// Here I create a second state for the menu drawer state (we can have multiple states with hooks)
// This is the equivalent to this.state.menuIsOpen and this.setState({ menuIsOpen })
const [menuIsOpen, setMenuState] = useState(false);
```

// Here I use a context to pass down navigation pages from the App.jsx
// The useContext hook is a hook coming from React to access data from a context without
// the use of any HoC
// inside we have navigationState.navPages and navigationState.drawerPages
const navigationState = useContext(navigationContext);

// Here I get the small value from redux
// useSelector is a hook from react-redux which allows us to select any value from the current redux state
const small = useSelector(state => state.browser.greaterThan.small);

// Same here, this gets the current user from redux
const user = useSelector(state => state.app.metadata.data.user);

// Same here, we get the apps strings
const strings = useSelector(state => state.app.strings);

bdbch commented on src/components/Header/Header.jsx:
```diff
@@ -139,105 +144,130 @@ const SettingsGroup = ({ children }) => {
+
+  useEffect(() => {
+    import('../Announce').then(ann => setAnnounce(ann.default));
+  }, []);
```
This is special. `useEffect` can be used to handle stuff like initialization, value updates etc.

This effect basically is a `componentDidMount` because it will only run once on the first render call and afterwards won't called again. Thats because we passed an empty dependency array to it (see line 222).

Inside the effect we set the announceState to the imported Announce component.

This would look like this in a class:

```js
componentDidMount() {
  import('../Announce').then(ann => this.setState({
    Announce: ann.default,
  }))
}
```

bdbch commented on src/components/Header/Header.jsx:
```diff
@@ -70,57 +74,58 @@ const DropdownMenuItem = styled(MenuItem)`
-};
-
 const SettingsGroup = ({ children }) => {
   const [anchorEl, setAnchorEl] = useState(null);
```
Here I set an anchorEl state which I will use to define were the dropdown menu should be attached.

bdbch commented on src/components/Header/Header.jsx:
```diff
@@ -70,57 +74,58 @@ const DropdownMenuItem = styled(MenuItem)`
-
 const SettingsGroup = ({ children }) => {
   const [anchorEl, setAnchorEl] = useState(null);
   const handleClose = useCallback(() => {
```
`useCallback` defines a callback function which will be called whenever handleClose is called. We need useCallback here so the function is not redefined everytime on render.

102 replied:
just a nitpick: arrow function inside `useCallback` will still be created on each render & `handleClose` will also be redefined on each render, but `handleClose` will keep consistent reference until dependencies array items changes

102 commented on package.json:
```diff
@@ -48,7 +48,7 @@
     "react-helmet": "^5.2.0",
     "react-markdown": "^4.0.3",
-    "react-redux": "^5.1.1",
+    "react-redux": "^7.1.1",
```
> refactored navbar styling and logo spacings 

I think we should do such upgrades via the separate PRs due to high chance of breaking something (2 major versions is not a joke)

howardchung replied:
I looked over the changelogs, doesn't seem to be anything breaking that we use: https://github.com/reduxjs/react-redux/releases

howardchung commented on src/components/App/App.jsx:
```diff
@@ -163,16 +165,10 @@ class App extends React.Component {
         to: '/combos',
```
I think this somehow got mixed up in a previous change. Combos and Teams are different paths, teams should go in the main navbar and combos in the drawer.

Also, I think we should add Explorer and API to the main navbar.

bdbch commented:
@howardchung should be fine now

howardchung commented on src/context/navigationContext.js:
```diff
@@ -0,0 +1,8 @@
+import { createContext } from 'react';
+
+const navigationContext = createContext({
+  navbarPages: [],
```
Why do these need to be in a context? Aren't they static and can be imported as constants?

bdbch replied:
We can't move these into easy to import functions without adding extra logic since we use the strings from our Redux store, thats why I'm defining them in a context so we can easily pass them down to any component below.

Otherwise we could export them from the constants file and add the strings key instead of the translated string.

howardchung replied:
The old version just passed the navbarPages down to the Header component. Perhaps we can use the same approach and pass two props?

bdbch replied:
I removed the context, see my latest commit.

howardchung commented:
Couple of lint errors, otherwise looks good

howardchung commented:
@bdbch are you able to fix the lint issue? If you're busy, I can do it!

bdbch commented:
@howardchung I can do it in a few minutes! :) I'll let you know when I updated my PR. thanks!

Argn0 commented:
there seems to be a bug

if you click on this link you land on the wrong tab

https://deploy-preview-2219--opendota.netlify.com/matches/5079685117/combat

bdbch commented:
Good find @Argn0 I'll try to fix it together with the lint issues now. Had a rough day today so I could just get back right now.

Seems to be an issue with the tab index.

bdbch commented:
@Argn0 I found the issue and fixed it with my last commit. The problem was that I was mapping through all tabs before filtering them and using the index to navigate/set active states of the tab. Since some tabs are not visible sometimes I moved the filtering out of the rendering block and moved it into a memorized variable.

bdbch commented:
@howardchung the PR should be fine now

howardchung commented:
I'm still seeing the issue with the link going to the wrong tab?

howardchung commented on src/components/TabBar/TabBar.jsx:
```diff
@@ -44,6 +45,8 @@ const TabBar = ({ tabs, match }) => {
     setTabValue(index);
   }, [history]);
 
+  const visibleTabs = useMemo(() => tabs.filter(tab => (!tab.hidden || (tab.hidden && !tab.hidden(match)))), [tabs]);
```
isn't the only change here moving the logic? I'm not sure why this would change the behavior.

howardchung replied:
Possibly an issue with the previous commit https://github.com/odota/web/commit/fc10a7cfc401a3f6727271a5d17b041fea331271#diff-571eda7df4f281922d74a5ed5b6a60c7?

bdbch replied:
Let me check again. For me it was working fine.

>isn't the only change here moving the logic? I'm not sure why this would change the behavior.

No it also changes the index while iterating through the list of tabs. That was what caused the issue in the first place since I used the "setActive" index of the unfiltered list, but the filtered list was rendered.

So the active tab was for example `3` while the actual tab index was `2` because one tab was not visible.

Moving it in front of the rendering map allows us to get the correct index. Let me check why it is not working

bdbch replied:
Ah so the problem is fixed but the issue with initially loading the correct tab is broken.
![fldCVHrxPe](https://user-images.githubusercontent.com/6538827/67610123-ffc59180-f790-11e9-9724-c493656de566.gif)

As you can see the tab is right when you actually click around.

But on initial load the tab is wrong:
![YCgskM1WGL](https://user-images.githubusercontent.com/6538827/67610141-1e2b8d00-f791-11e9-8df0-2944209cf9d8.gif)

I'll fix the problem!

bdbch replied:
Found the problem. A few lines above in `TabBar.jsx` I used `useEffect` to set the initial tab index without using the filtered tabs list.

Changing it to this:

```jsx
useEffect(() => {
  visibleTabs.forEach((tab, i) => {
    if (location.pathname === tab.route) setTabValue(i);
  });
}, [visibleTabs]);
```

worked fine:
![ARJom5Ey3k](https://user-images.githubusercontent.com/6538827/67610212-811d2400-f791-11e9-9b60-1a9bc6720bdb.gif)

Argn0 commented:
> I'm still seeing the issue with the link going to the wrong tab?

me too

bdbch commented:
> > I'm still seeing the issue with the link going to the wrong tab?
> 
> me too

could you retest this with the latest netlify deploy preview?

Argn0 commented:
> > > I'm still seeing the issue with the link going to the wrong tab?
> > 
> > 
> > me too
> 
> could you retest this with the latest netlify deploy preview?

It seems to work correctly now.

bdbch commented:
@howardchung can we merge this pr?
